plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+
+            CompletableFuture<CompletedPart> uploadFuture = uploadPart(uploadId.get(), slice);
+            uploadFutures.add(uploadFuture);
+
+            return asVoid(toListenableFuture(uploadFuture));
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            if (uploadId.isEmpty()) {
+                return;
+            }
+
+            try {
+                List<CompletedPart> completedParts = uploadFutures.stream()
+                        .map(CompletableFuture::join)
+                        .sorted(Comparator.comparing(CompletedPart::partNumber))
+                        .collect(toImmutableList());
+                completeMultiUpload(uploadId.get(), completedParts);
+            }
+            catch (RuntimeException e) {
+                abortUploadSuppressed(uploadId.get(), e);
+                throw new IOException(e);
+            }
+        }
+
+        private CreateMultipartUploadResponse createMultipartUpload()
+        {
+            CreateMultipartUploadRequest.Builder createMultipartUploadRequestBuilder = CreateMultipartUploadRequest.builder()
+                    .bucket(bucketName)
+                    .key(key);
+            S3RequestUtil.configureEncryption(secretKey, createMultipartUploadRequestBuilder);
+            return s3Client.createMultipartUpload(createMultipartUploadRequestBuilder.build());
+        }
+
+        private CompletableFuture<CompletedPart> uploadPart(String uploadId, Slice slice)
+        {
+            currentPartNumber++;
+            UploadPartRequest.Builder uploadPartRequestBuilder = UploadPartRequest.builder()
+                    .bucket(bucketName)
+                    .key(key)
+                    .uploadId(uploadId)
+                    .partNumber(currentPartNumber);
+            S3RequestUtil.configureEncryption(secretKey, uploadPartRequestBuilder);
+            UploadPartRequest uploadPartRequest = uploadPartRequestBuilder.build();
+            return s3AsyncClient.uploadPart(uploadPartRequest, DirectByteArrayAsyncRequestBody.fromByteBuffer(slice.toByteBuffer()))
+                    .thenApply(uploadPartResponse -> CompletedPart.builder().eTag(uploadPartResponse.eTag()).partNumber(uploadPartRequest.partNumber()).build());
+        }
+
+        private void completeMultiUpload(String uploadId, List<CompletedPart> completedParts)
+        {
+            CompletedMultipartUpload completedMultipartUpload = CompletedMultipartUpload.builder()
+                    .parts(completedParts)
+                    .build();
+            CompleteMultipartUploadRequest completeMultipartUploadRequest = CompleteMultipartUploadRequest.builder()
+                    .bucket(bucketName)
+                    .key(key)
+                    .uploadId(uploadId)
+                    .multipartUpload(completedMultipartUpload)
+                    .build();
+            s3Client.completeMultipartUpload(completeMultipartUploadRequest);
```

Comment:
Same here

Reply:
This essentially ties to the comments above --- whether `ExchangeSink#finish` and `ExchangeSink#abort` should be asynchronous. If not I feel that we have to do this synchronously to guarantee semantic correctness.